tools/ocr.py

```diff
@@ -0,0 +1,275 @@
+                continue
+            recog_results = model_inference(
+                recog_model, chunk_box_imgs, batch_mode=True)
+            for i, recog_result in enumerate(recog_results):
+                text = recog_result['text']
+                text_score = recog_result['score']
+                if isinstance(text_score, list):
+                    text_score = sum(text_score) / max(1, len(text))
+                end2end_res['result'][start_idx + i]['text'] = text
+                end2end_res['result'][start_idx + i]['text_score'] = text_score
+
+    return end2end_res
+
+
+def main():
+    args = parse_args()
+    ocr = MMOCR(**vars(args))
+    ocr.readtext(**vars(args))
+
+
+def parse_args():
+    parser = ArgumentParser()
+    parser.add_argument('img', type=str, help='Input Image file.')
+    parser.add_argument('--out_img', type=str, default='',
+                        help='Output file name of the visualized image.')
+    parser.add_argument('--textdet', type=str,
+                        default='DB_r18', help='Text detection algorithm')
+    parser.add_argument('--textrecog', type=str,
+                        default='CRNN', help='Text recognition algorithm')
+    parser.add_argument('--batch-mode', action='store_true',
+                        help='Whether use batch mode for text recognition.')
+    parser.add_argument('--batch-size', type=int, default=4,
+                        help='Batch size for text recognition inference if batch_mode is True above.')
+    parser.add_argument('--device', default='cuda:0',
+                        help='Device used for inference.')
+    parser.add_argument('--export-json', action='store_true',
+                        help='Whether export the ocr results in a json file.')
+    parser.add_argument('--details', action='store_true',
+                        help='Whether include the text boxes coordinates and confidence values')
+    parser.add_argument('--imshow', action='store_true',
+                        help='Whether show image with OpenCV.')
+    parser.add_argument('--ocr-in-lines', action='store_true',
+                        help='Whether group ocr results in lines.')
+    parser.add_argument('--print-result', action='store_true',
+                        help='Prints the recognised text')
+    args = parser.parse_args()
+    return args
+
+
+class MMOCR:
+    def __init__(self, textdet='DB_r18', textrecog='CRNN', device='cuda:0', **kwargs):
+        self.td = textdet
+        self.tr = textrecog
+        if device == 'cpu':
+            self.device = 0
+        else:
+            self.device = device
+
+        if self.td not in textdet_models:
+            raise ValueError(
+                self.td, 'is not a supported text detection algorthm')
+        elif self.tr not in textrecog_models:
+            raise ValueError(
+                self.tr, 'is not a supported text recognition algorithm')
+
+        dir_path = os.path.dirname(os.path.realpath(__file__))
```

Comment:
Use `dir_path = os.getcwd()`. We always assume user runs the script like
```
python tools/ocr.py demo/demo_text_det.jpg --imshow
```

Reply:
Wouldn't it be more flexible if we get the absolute path of the config files based on the location of the ocr.py file ? So that the user can run the script from whatever cwd directory he's in. Or will that break other things ? 
Also, why not put the file in the mmocr/utils directory so that the user can call it from any python script with :

```python
from mmocr.utils.ocr import MMOCR
```

Additionally, I believe the config folder should maybe be included in the installation setup (currently excluded). Maybe you can shine some light on the structural choices that have been made ? Because I don't see another way to make mmocr run easily out-of-the-box without extra steps.

Reply:
> Wouldn't it be more flexible if we get the absolute path of the config files based on the location of the ocr.py file ? So that the user can run the script from whatever cwd directory he's in. Or will that break other things?
> ...
> Additionally, I believe the config folder should maybe be included in the installation setup (currently excluded). Maybe you can shine some light on the structural choices that have been made? Because I don't see another way to make mmocr run easily out-of-the-box without extra steps.

Including it into the package is not suggested because config files are more flexible than code and user should be able to modify them. However, it is possible to get the relative path of config files given your script file location. (btw, it might also be cool to support customized config path in case users want to test their own models)

> Also, why not put the file in the mmocr/utils directory so that the user can call it from any python script with :
> 
> ```python
> from mmocr.utils.ocr import MMOCR
> ```
Good idea. Also don't forget to modify `mmocr/utils/__init__.py` so that `from mmocr.utils import MMOCR` also works.

Reply:
I moved the file to `mmocr/utils`, modified the `mmocr/utils/__init__.py` and added arguments for custom detection and recognition configs. If there's anything else you think I should add as an argument let me know. Otherwise I will start writing something for the docs.